clients/drcachesim/tracer/func_trace.cpp

```diff
@@ -101,16 +105,20 @@ func_pre_hook(void *wrapcxt, INOUT void **user_data)
     if (drcontext == NULL)
         return;
 
+    void *pt_data = drmgr_get_tls_field(drcontext, tls_idx);
+    func_trace_entry_vector_t *vec = (func_trace_entry_vector_t *)pt_data;
+    vec->size = 0;
     size_t idx = (size_t)*user_data;
     func_metadata_t *f = (func_metadata_t *)drvector_get_entry(&funcs, (uint)idx);
     app_pc retaddr = drwrap_get_retaddr(wrapcxt);
-    append_entry(drcontext, TRACE_MARKER_TYPE_FUNC_ID, (uintptr_t)f->id);
-    append_entry(drcontext, TRACE_MARKER_TYPE_FUNC_RETADDR, (uintptr_t)retaddr);
+
+    ADD_TO_VEC(vec, TRACE_MARKER_TYPE_FUNC_ID, (uintptr_t)f->id);
+    ADD_TO_VEC(vec, TRACE_MARKER_TYPE_FUNC_RETADDR, (uintptr_t)retaddr);
     for (int i = 0; i < f->arg_num; i++) {
         uintptr_t arg_i = (uintptr_t)drwrap_get_arg(wrapcxt, i);
-        append_entry(drcontext, TRACE_MARKER_TYPE_FUNC_ARG, arg_i);
+        ADD_TO_VEC(vec, TRACE_MARKER_TYPE_FUNC_ARG, arg_i);
     }
-    memtrace_if_redzone(drcontext);
+    append_entry_vec(drcontext, vec);
```

Comment:
Hmm, if the indirect call is an overhead concern, maybe we should try to turn it into a direct inline-able call and keep separate calls, rather than adding memory pressure by having to store into this array.

It would also be good to profile this on some kind of benchmark so we have numbers to show this is doing what we think it is.

Given that this seems to be a (potentially short-term) tweak to try and get large app overhead under a threshold, I'm ok not spending much time on it (i.e., going with this array interface for now) if we add a comment explaining that there may be better ways to do this and that we should profile and improve this at some point.

Reply:
I will put down the detailed comment about the reason behind and future improve point.